training/architecture/12_FACTOR_APP_PRINCIPLES.md

```diff
@@ -0,0 +1,167 @@
+
+We aim to provide open source software; if we were to hard code our configuration in the service, we might disclose information which could raise security issues. In order to protect information, it is good practice to make sure any secrets, passwords are set using config and never hard coded into an app.
+
+The configuration for each app exists in one place so that we can collaborate and manage them easily, for example: `config/config.go` in our go applications.
+
+In our Go apps we use the [envconfig package](https://github.com/kelseyhightower/envconfig) to help us implement this principle.
+
+
+### 4. Backing services
+Pre-requisite: [Backing services](https://12factor.net/backing-services)
+
+> Treat backing services as attached resources
+
+Our apps rely on the use of external services, such as databases (e.g. MongoDB).
+A 12-factor app should be able to switch out these external services by modifying the configuration, e.g. with no code changes.
+
+### 5. Build, release, run
+Pre-requisite: [Build, release, run](https://12factor.net/build-release-run)
+
+> Strictly separate build and run stages
+
+We have created pipelines that build, release and run our apps.
+1. Build: based on a commit, the code is converted into an executable bundle i.e. a build.
+2. Release: the [dp-configs](https://github.com/ONSdigital/dp-configs) configuration is applied to the build from the previous stage.
+3. Run: the app is deployed to the environment (`develop` or `production`) using the selected release.
+
+Strict separation means that any change has to go through the stages, rather than being applied directly at runtime.
+
+We use [Semantic Versioning](https://semver.org/) to give each release a unique id. This allows us to easily rollback to earlier releases.
+
+Find out more:
+- [DP Concourse documentation](https://github.com/ONSdigital/dp/blob/master/training/platform-services/PLATFORM.md#concourse)
+- Head to [concourse.onsdigital.co.uk/](https://concourse.onsdigital.co.uk/) to see our pipelines.
+
+
+### 6. Processes
+Pre-requisite: [Processes](https://12factor.net/processes)
+
+> Execute the app as one or more stateless processes
+ 
+A 12-factor app is stateless, i.e. it does not save client data from a session to be used in the next session with that client. This means that restarting the app does not result in different outcomes when processing information or requests.
+
+In DP, the data that persist is stored in stateful backing service (database... or static file?)
+
```

Comment:
Should this be a question?

Reply:
I'm just not sure what I'm saying makes sense 🙈

Reply:
have added a suggestion